text/0000-plugin-package-format.md

```diff
@@ -0,0 +1,111 @@
+
+# Drawbacks
+Potentially more complicated than current logic.
+
+# Design
+All plugins should be shipped and loaded in the following format:
+
+- `/<name>/data`
+    The plugins custom data, such as localized text and images.
+- `/<name>/bin/<platform>/[.../]<name>[.<ext>]`
+    Binaries for a specific platform. Platforms may choose to do additional parsing of the path in order to support version selection, architectures, and bitness.
+
+## Keywords
+### Keyword: `<name>`
+The unique name of the plugin which must match the actual module name, excluding the file extension. This name can be anything, as long as it is valid for the platform and/or container the plugin is in.
+
+### Optional Keyword: `<ext>`
+A platform specific extension to denote dynamic libaries. A generic extension, such as `bin`, `obs` or similar may also be supported, depending on the final implementation, but this RFC does not specify that anything but the platform native extension for dynamic libraries. On platforms where no such extension is required, the dot prefixing the extension is also removed.
+
+### Keyword: `<platform>`
+The platform for which the binary is made for. Platforms may opt for additional parsing of the `[.../]` part in order to filter by platform version, architecture or bitness. By default the following platforms are available:
+
+#### `windows`
+The Windows platform supports the additional keywords `<version>`, `<arch>` and `<bits>`, which should be written as: 
+
+```
+/<name>/bin/windows[/<version>]/<arch>-<bits>/<name>[.<ext>]
+```
+
+- `<version>` is optional.
+- `<arch>` is required.
+- `<bits>` is required.
+
+This layout allows us to (optionally) specify a minimum platform version, a required architecture, and a required bitness. For example, with this we can filter for Windows 10 20H2 x86 64-bits with the following text: `/<name>/bin/windows/10.0.19042/x86-64/<name>[.<ext>]`, and as such prevent the plugin from loading on unsupported platforms automatically.
+
+#### `mac` or `macos`
+The Mac/MacOS platform supports the additional keywords `<version>`, `<arch>` and `<bits>`, which should be written as:
+
+```
+/<name>/bin/mac[/<version>]/<arch>-<bits>/<name>[.<ext>]
+```
+
+- `<version>` is optional.
+- `<arch>` is required.
+- `<bits>` is required.
+
+This layout allows us to (optionally) specify a minimum platform version, an required architecture, and a required bitness. We can easily filter for specific Apple devices, and require certain OS versions for plugins, or even have M1 or X86 exclusive plugins.
+
+#### `linux`
+The Linux platform supports the additional keywords `<distro>`, `<version>` `<arch>` and `<bits>`, which should be written as:
+
+```
+/<name>/bin/linux[/<distro>[-<version>]]/<arch>-<bits>/<name>[.<ext>]
+```
+
+- `<distro>` is optional, and specific distributions may have a fallback to other values if no specific version is present.
+- `<version>` is optional.
```

Comment:
I notice there's no elaboration on what "version" means on Linux. As every single distro will be different, maybe we should (as an example) include how it'd be done on Ubuntu, and when support for other distros is added, there is a basis to work off.

It's important to note that for Ubuntu currently, we [build a separate version](https://launchpad.net/~obsproject/+archive/ubuntu/obs-studio/+packages) of OBS for each release of Ubuntu as the OS ships with different versions of our dependencies.

Reply:
Yes, I could not find a unified cross-distro way to identify what the platform version is to really define it. On Ubuntu and Debian, I found that the command `lsb_release -a` gives a reasonable result, though Debian for whatever reason returns text instead of version numbers. This needs input from people that actually know how to identify a platform version.

Reply:
Seems like Debian is the only(?) outlier here with not giving a proper version and patch level.

Reply:
I've added a better definition of the version keyword for linux now.